gaps/gap-4_comp_manifest/gap-4_comp_manifest.md

```diff
@@ -0,0 +1,278 @@
+  Golem Computation Resource Standards.
+
+The Computation Manifest is expected to be included in a Computation Payload
+Manifest ([GAP](https://github.com/golemfactory/golem-architecture/pull/28))
+file, where its hash is co-signed with Payload hashes. If a Computation
+Manifest is included in the Demand (and thus, Agreement), its role is to
+derive the properties from and override the initially sourced Manifest.
+
+### GCRS `golem.srv.comp.manifest` namespace
+
+Intended for overriding an external Manifest but allows defining a new Manifest
+from the ground up using properties only.
+
+1. `golem.srv.comp.manifest.version : String`
+
+    Specifies a version (Semantic Versioning 2.0 specification) of the manifest,
+    **defaults** to "0.1.0"
+
+2. `golem.srv.comp.manifest.script` (sub-namespace)
+
+    Defines a set of allowed ExeScript commands and applies constraints to
+    their arguments.
+
+
+3. `golem.srv.comp.manifest.script.commands : List[String]`
+
+    Specifies a curated list of commands in form of:
+
+    - UTF-8 encoded strings
+
+      No command context or matching mode need to be specified.
+
+      E.g. `["run /bin/cat /etc/motd", "run /bin/date -R"]`
+
+    - UTF-8 encoded JSON strings
+
+      Command context (e.g. `env`) or argument matching mode need to be
+      specified for a command.
+
+      E.g. `["{\"run\": { \"args\": \"/bin/date -R\", \"env\": { \"MYVAR\": \"42\", \"match\": \"strict\" }}}"]`
+
+    - mix of both
+
+    `"deploy"`, `"start"` and `"terminate'` commands are always allowed.
+    These values become the **default** if no `manifest.script.command` property
+    has been set in the Demand, but the `manifest` namespace is present.
+
+4. `golem.srv.comp.manifest.script.match : String`
+
+    Selects a default way of comparing command arguments stated in the manifest
+    and the ones received in the ExeScript, unless stated otherwise in a
+    command JSON object.
+
+    The `match` property could be one of:
+
+      - `"strict"`: byte-to-byte argument equality (**default**)
+      - `"regex"`: treat arguments as regular expressions
+
+        Syntax: Perl-compatible regular expressions (UTF-8 Unicode mode),
+        w/o the support for look around and backreferences (among others);
+        for more information read the documentation of the Rust
+        [regex](https://docs.rs/regex/latest/regex/) crate.
+
+5. `golem.srv.comp.manifest.net` (sub-namespace)
+
+    Applies constraints to networking. Currently, outgoing requests to the
+    public Internet network are covered.
+
+6. `golem.srv.comp.manifest.net.inet.out.protocols : List[String]`
+
+    List of allowed outbound protocols. Currently **fixed at** `["http", "https"]`.
+
+7. `golem.srv.comp.manifest.net.inet.out.urls : List[String]`
+
+    List of allowed external URLs that outbound requests can be sent to.
+    E.g. `["http://golemfactory.s3.amazonaws.com/file1", "http://golemfactory.s3.amazonaws.com/file2"]`
```

Comment:
how would we verify a URL on a network-interface level? shouldn't it rather be an IP address? or an IP range?

Reply:
This question goes well beyond the scope of this GAP :slightly_smiling_face: what you're asking of would be a separate property, not covered here (`net.inet.out.ips` ?).

As of now, filtering is done on OSI L3+ layers (IP and upwards), and domain names have to be resolved (to an address or a range of addresses) before we start executing commands. 

The upsides:
- IP addresses of servers specified in the manifest can change
- developers use human-readable names for destinations

The downsides:
- ExeUnit starts with a delay for DNS name resolution, w/ potential TLS certificate issues
- DNS records can change during execution, so we need a background job to update those (in the future)